modules/standard/Types.chpl

```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
 pragma "no doc"
 inline proc _bxor_id(type t) return 0:t;
 
+
+/* Return the number of fields in a class or record as a param.
+   The count of fields includes types and param fields.
+ */
+proc numFields(type t) param
+  return __primitive("num fields", t);
+
+/* Get the name of the `i`th field in a class or record.
+   `i` should be in 1..numFields(t).
+
+   :arg t: a class or record type
+   :arg i: which field to get the name of
+   :returns: the name of the field, as a param string
+ */
+proc getFieldName(type t, param i:int) param
+  return __primitive("field num to name", t, i);
+
+pragma "no doc"
+proc fieldNumToName(type t, param i:int) param {
+  compilerError("fieldNumToName deprecated. Use getFieldName");
+}
+
+/* Get the `i`th field in a class or record.
+   `i` should be in 1..numFields(t).
+
+   :arg x: a class or record
+   :arg i: which field to get
+   :returns: an rvalue referring to that field.
+ */
+inline
+proc getField(x:?t, param i:int)
+  return __primitive("field value by num", x, i);
+
+// TODO ref return version of getField(x, int)
+
+proc fieldValueByNum(x, param i:int) {
+  compilerError("fieldValueByNum deprecated. Use getField");
+}
+
+/* Get a field in a class or record by name.
+   Will generate a compilation error if a field with that name
+   is not found.
+
+   :arg x: a class or record
+   :arg s: the name of a field
+   :returns: an rvalue referring to that field.
+ */
+proc getField(x:?t, param s:string) {
+  param i = __primitive("field name to num", t, s);
+  if i == 0 then
+    compilerError("field ", s, " not found in ", t:string);
+  return __primitive("field value by num", x, i);
+}
+
+// TODO ref return version of getField(x, string)
+
+/* Get a field index in a class or record.
+   Will not generate a compilation error if the field
+   is not found.
+
+   :arg t: a class or record type
+   :arg s: the name of a field
+   :returns: an index `i` usable in getField, or 0 if the field was not found.
+ */
+proc getFieldIndex(type t, param s:string) param
+  return __primitive("field name to num", t, s);
+
+/* Returns `true` if a class or record has a field named `s`,
+   or `false` otherwise.
+
+   :arg t: a class or record type
+   :arg s: the name of a field
+   :returns: `true` if the field is present.
+ */
+
+proc hasField(type t, param s:string) param
+  return getFieldIndex(t, s) > 0;
+
+/* Returns true if a function named `fname` taking in no arguments
+   is available in the current scope.
+   */
+proc canResolve(param fname) param
+  return __primitive("call resolves", fname);
+
+/* Returns true if a function named `fname` taking the arguments in args
+   is available in the current scope.
+   */
+proc canResolve(param fname, args ...?k) param
+  return __primitive("call resolves", fname, (...args));
+
+// TODO -- how can this work with by-name argument passing?
+
+/* Returns true if a method on x named `fname` taking in no arguments
+   is available in the current scope.
+   */
```

Comment:
'obj' or 'object' seems pretty good.  I don't like the underscore in '_this'.  I suppose there's even a chance that 'this' might work (we don't reserve it I don't think, do we?) but that seems like it might be confusing.  This conversation makes me think that this is another routine that would, ideally, be great to have be a type method on the object itself (acknowledging that there are challenges there).